### PR TITLE
api_wraps: Preserve name of original wrapper function (fixes #296)

### LINKED ISSUE #296: Names of actions added with 'wrap' should be displayed in '--seneca.print.tree'
I have a chain of 'interceptors' applied like so:
```javascript
// index.js

var seneca = require('seneca')()
seneca.use('math').listen()
```

```javascript
// math.js

module.exports = function (options) {
  this.add('role:math, cmd:sum', function sum (msg, respond) {
    const answer = msg.left + msg.right
    respond(null, {answer})
  })

  this.add('role:math, cmd:product', function product (msg, respond) {
    const answer = msg.left * msg.right
    respond(null, {answer})
  })

  this.wrap('role:math', function integerSupport (msg, respond) {
    if (!msg.integer) {
      return this.prior(msg, respond)
    }

    if ('left' in msg) {
      msg.left = Math.floor(msg.left)
    }
    if ('right' in msg) {
      msg.right = Math.floor(msg.right)
    }
    this.prior(msg, respond)
  })

  this.wrap('role:math', function finiteOnly (msg, respond) {
    if (!Number.isFinite(msg.left) || !Number.isFinite(msg.right)) {
      respond(new Error('Expected left and right to be numbers.'))
    } else {
      this.prior(msg, respond)
    }
  })

  this.wrap('role:math', function logOperations (msg, respond) {
    this.make('operation', {cmd: msg.cmd, left: msg.left, right: msg.right}).save$()
    this.prior(msg, respond)
  })

  this.wrap('role:math', function numberify (msg, respond) {
    if ('left' in msg) {
      msg.left = Number(msg.left)
    }
    if ('right' in msg) {
      msg.right = Number(msg.right)
    }
    this.prior(msg, respond)
  })
}
```
Note all functions are named (and in fact their 'name' property has correct value).
But when I do `node index.js --seneca.print.tree` only actions added by `add` are shown.
```
Seneca action patterns for instance: rh2pbps0my5q/1452205680107/11579/-
├─┬ cmd:sum
│ └─┬ role:math
│   └── # math, (q67tw), 
│       # math, (5um5m), 
│       # math, (hz0v7), 
│       # math, (nmbqq), 
│       # math, (65qzr), sum
└─┬ cmd:product
  └─┬ role:math
    └── # math, (dhvcw), 
        # math, (d5rbj), 
        # math, (wxbgw), 
        # math, (z10ot), 
        # math, (x4yzb), product
```
Looking at the code the cause seems to be that in`api_wrap` passed to api_add is not the original function from user but an anonymous wrapper that does not preserve the function `name` property (which is what is actual output in the tree)

A possible solution would be to assign some hidden property to the func object itself and that would be used in print.js instead of the current `meta.func.name` or to actually "fix" the name of the anonymous wrapper. 

Since javascript does not allow direct modification of functions' `name` property I stumbled upon solution to rename functions like this
```javascript
function renameFunction (name, fn) {
  return (new Function("return function (call) { return function " + name +
    "  () { return call(this, arguments) }; };")())(Function.apply.bind(fn));
}
```
But running a simple micro-benchmark, it seems to add quite significant performance overhead
```
basic x 113,936,140 ops/sec ±2.80% (86 runs sampled)
renamed x 3,589,117 ops/sec ±0.99% (85 runs sampled)
```

Other solution to rename the function properly (without the performance bottleneck) would be to use 
```javascript 
Object.defineProperty(fn, 'name', {value: desiredName})
```
but this appears to work only in Node 4 and above.

Though with both above solutions I managed to get correct output
```
Seneca action patterns for instance: wm9ibot6tir5/1452207249487/14112/-
├─┬ cmd:sum
│ └─┬ role:math
│   └── # math, (il09y), numberify
│       # math, (22zru), logOperations
│       # math, (i5jjo), finiteOnly
│       # math, (5rluy), integerSupport
│       # math, (qcy7c), sum
└─┬ cmd:product
  └─┬ role:math
    └── # math, (te5c8), numberify
        # math, (pudnl), logOperations
        # math, (so7ow), finiteOnly
        # math, (y9uuz), integerSupport
        # math, (6pp0t), product
```

Since you seem to support all the way down to 0.10 I will try to make PR with the `Object.defineProperty` solution with graceful degradation on older Nodes. 

Though regardless of the final solution I think it is quite useful to see the actual chain of what functions are actually applied and it really helped me with debugging.

### PR DESCRIPTION
**Pull request by matobet**
Fixes #296

### CODE REVIEW

File: seneca.js [hunk 1]
```diff
@@ -891,9 +891,7 @@ function make_seneca (initial_options) {
     pin = _.isArray(pin) ? pin : [pin]
     _.each(pin, function (p) {
       _.each( pinthis.findpins(p), function (actpattern) {
-        pinthis.add(actpattern, meta, function (args, done) {
-          wrapper.call(this, args, done)
-        })
+        pinthis.add(actpattern, meta, wrapper)
       })
```
Comment (mcdonnelldean) [inline seneca.js]: Are we not losing the explicit bound `this` in this change?

Comment (matobet) [inline seneca.js]: Actually what the outer `function` does is to add another scope of `this` which we are getting rid of by the `call` so the `wrapper` would be called with the same this as specified by `add`. When you think about it this is exactly what happens when you pass any function to `add` and it is also true for user supplied `wrapper`.

In my sample code in #296 I'm using the `this` fom inside the wrappers and they still work correctly.

Comment (mcdonnelldean) [inline seneca.js]: That's what I was thinking but I want to be sure there wasn't an explicit reason for this code. @geek Can you take a quick look?

**Comment by geek** [comment]
@matobet can you add a test to demonstrate the fix

**Comment by matobet** [comment]
@geek sure, posted new version

**Pull request merged**